aten/src/ATen/native/miopen/Conv.cpp

```diff
@@ -0,0 +1,988 @@
+                        - kernel) / stride[d - 2] + 1;
+  }
+  return output_size;
+}
+
+std::vector<int64_t> conv_input_size(
+    IntList output_size, IntList weight_size,
+    IntList padding, IntList output_padding, IntList stride, IntList dilation, int64_t groups
+) {
+  // ASSERT(output_size.size() > 2)
+  // ASSERT(output_size.size() == weight_size.size())
+  auto dim = output_size.size();
+  std::vector<int64_t> input_size(dim);
+  input_size[0] = output_size[output_batch_size_dim];
+  input_size[1] = weight_size[weight_input_channels_dim] * groups;
+  for (size_t d = 2; d < dim; ++d) {
+    int kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
+    input_size[d] = (output_size[d] - 1) * stride[d - 2] - (2 * padding[d - 2]) +
+                     kernel + output_padding[d - 2];
+  }
+  return input_size;
+}
+
+std::vector<int64_t> conv_weight_size(
+    IntList input_size, IntList output_size,
+    IntList padding, IntList output_padding, IntList stride, IntList dilation, int64_t groups
+) {
+  auto dim = input_size.size();
+  std::vector<int64_t> weight_size(dim);
+  weight_size[0] = output_size[1];
+  weight_size[1] = input_size[1] / groups;
+  for (size_t d = 2; d < dim; ++d) {
+    int kernel = input_size[d] - (output_size[d] - 1) * stride[d - 2]
+               + 2 * padding[d - 2] - output_padding[d - 2];
+    weight_size[d] = (kernel - 1) / dilation[d - 2] + 1;
+  }
+  return weight_size;
+}
+
+Tensor narrowGroup(const Tensor& t, int dim, int group_idx, int64_t groups) {
+  auto group_size = t.size(dim) / groups;
+  return t.narrow(dim, group_idx * group_size, group_size);
+}
+
+// ---------------------------------------------------------------------
+//
+// Checking
+//
+// ---------------------------------------------------------------------
+
+// Used on pad, stride and dilation
+static void check_args(CheckedFrom c, IntList args, size_t expected_size, const char* arg_name)
+{
+  if (args.size() > expected_size){
+    std::stringstream ss;
+    ss << "Too many " << arg_name << " values (" << args.size() << ") supplied, expecting " << expected_size << " (while checking arguments for " << c << ")";
+    throw std::runtime_error(ss.str());
+  }
+  else if (args.size() < expected_size){
+    std::stringstream ss;
+    ss << "Not enough " << arg_name << " values (" << args.size() << ") supplied, expecting " << expected_size << " (while checking arguments for " << c << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  auto num_negative_values = std::count_if(args.begin(), args.end(), [](int x){return x < 0;});
+  if (num_negative_values > 0){
+    std::stringstream ss;
+    ss << arg_name << " should be greater than zero but got (";
+    std::copy(args.begin(), args.end() - 1, std::ostream_iterator<int>(ss,", "));
+    ss << args.back() <<  ")" << " (while checking arguments for " << c << ")";
+    throw std::runtime_error(ss.str());
+  }
+}
+
+
+// NB: For many call sites, it is not strictly necessary to check all of
```

Comment:
Can you turn the original version of this comment into a Note and reference it from here.